### PR TITLE
Derive probability for broadcasting operations

### PR DESCRIPTION
**Pull request by ricardoV94**
Related to #6398

TODO: 
- [ ] Cover `Second`/`Alloc` which are other froms of broadcasting

----
:books: Documentation preview :books:: https://pymc--6808.org.readthedocs.build/en/6808/

CC @shreyas3156

### CODE REVIEW

**Comment by larryshamalama** [comment]
I will get to this tomorrow morning, sorry about the delay!

File: pymc/logprob/shape.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+
+MeasurableVariable.register(MeasurableBroadcast)
+
+
+measurable_broadcast = MeasurableBroadcast()
+
+
+@_logprob.register(MeasurableBroadcast)
+def broadcast_logprob(op, values, rv, *shape, **kwargs):
```
Comment (larryshamalama) [inline pymc/logprob/shape.py:34]: Thinking out loud: could this possibly result in inconsistencies elsewhere? For instance, having Mixture components that have been broadcasted which would render them dependent, if that would be an issue

File: pymc/logprob/shape.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    ``logp(broadcast_to(normal(size=(1,)), (3,)), [1, 2, 3]) == [-np.inf]``
+
+    """
+    [value] = values
+
+    n_new_dims = len(shape) - rv.ndim
+    assert n_new_dims >= 0
+
+    # Enumerate broadcasted dims
```
Comment (larryshamalama) [inline pymc/logprob/shape.py:55]: Trying to follow along here, this comment is more for "mental scribbles".
```python
rv = pt.random.normal(size=(3, 1))
x = pt.broadcast_to(rv, (5, 2, 3, 4)) # a bit more than your example above
# rv.broadcastable = (False, False, False, False)

n_new_dims = 2 # 4 - 2
expanded_dims = (0, 1)

value.broadcastable[n_new_dims:] = (False, False) # (3, 4)
rv.broadcastable = (False, True) # (3, 1)

# condition is True only: if (not v_bcast) and rv_bcast = if (not False) and True
# condition is True only if v_bast is False and rv_bcast is True
broadcast_dims = (3,) # (0 + 2, 1 + 2) but conditions are (False, True)?
```

File: pymc/logprob/shape.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    expanded_dims = tuple(range(n_new_dims))
+    broadcast_dims = tuple(
+        i + n_new_dims
+        for i, (v_bcast, rv_bcast) in enumerate(
+            zip(value.broadcastable[n_new_dims:], rv.broadcastable)
+        )
+        if (not v_bcast) and rv_bcast
+    )
```
Comment (larryshamalama) [inline pymc/logprob/shape.py:63]: What is `broadcast_dims` in words? It's a bit hard to follow 😅

Comment (larryshamalama) [reply pymc/logprob/shape.py:63]: AH, is it the axis/es which is/are being broadcasted?

File: pymc/logprob/shape.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        return None  # pragma: no cover
+
+    rv_map_feature: Optional[PreserveRVMappings] = getattr(fgraph, "preserve_rv_mappings", None)
+
+    if rv_map_feature is None:
+        return None  # pragma: no cover
+
+    base_rv, *shape = node.inputs
+
+    if not rv_map_feature.request_measurable([base_rv]):
```
Comment (larryshamalama) [inline pymc/logprob/shape.py]: PR #6746 has been a while, but I just want to clarify `request_measurable`. If `base_rv` is unmeasurable, we exit? This method is the new `assign_custom_measurable_outputs`, is that right?

File: pymc/logprob/shape.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            indices.append(slice(None))
+
+    unbroadcast_value = value[tuple(indices)]
+    logp = _logprob_helper(rv, unbroadcast_value)
+
+    # Check that dependent values were indeed identical, by comparing with a re-broadcasted value
+    valid_value = pt.broadcast_to(unbroadcast_value, shape)
+    # Note: This could fail due to float-precision issues.
+    # If that proves to be a problem we should switch to `pt.allclose`
+    check = pt.all(pt.eq(value, valid_value))
```
Comment (larryshamalama) [inline pymc/logprob/shape.py]: It is my understanding that this segment (and some lines above) are to ensure that we have:
```python
logp(broadcast_to(normal(size=(1,)), (3,)), [1, 2, 3]) == [-np.inf]
```
Can you briefly explain the process of unbroadcasting and re-broadcasting to perform this check?

File: tests/logprob/test_transforms.py [hunk 6]
```diff
@@ -807,16 +807,36 @@ def test_discrete_rv_multinary_transform_fails():
         conditional_logp({y_rv: y_rv.clone()})
 
 
-@pytest.mark.xfail(reason="Check not implemented yet")
-def test_invalid_broadcasted_transform_rv_fails():
+@pytest.mark.filterwarnings("error")  # Fail if unexpected warning is issued
+@pytest.mark.parametrize("implicit_broadcast", (True, False))
+def test_broadcasted_transform_rv(implicit_broadcast):
```
Comment (larryshamalama) [inline tests/logprob/test_transforms.py]: Correct me if I'm wrong here: logprob inference for broadcasted variables work only if they are 1) pure RVs and 2) affine-transformed pure RVs?

**Review by larryshamalama — COMMENTED** [review]
Great work @ricardoV94! :) A lot of nice abstractions, which, together, are why I have many questions

File: pymc/logprob/shape.py (hunk 1 above)
Comment (ricardoV94) [reply pymc/logprob/shape.py:34]: The index mixture only works for basic RVs still so that's fine.

The switch mixture could actually wrongly broadcast the logp. In fact we should also check for invalid switches that mix support dimensions. The current implementation is only correct for `ndim_supp==0`!

This is another example of why it's so important to have the meta-info for all the MeasurableOps (#6360).

Once we have the meta-info, the Mixture will unambiguously know what kind of measurable variable it is dealing with. In the case of MeasurableBroadcasting, for example, the `ndim_supp` will have to be at least as large as the number of broadcasted dims (which means we should collapse that logp dimension instead of leaving it as we were doing now!). 

We will also know where those support dims are, so that Mixture can know whether we are sub-selecting across core dims.

Without the meta-info, the only way of knowing `ndim_supp` is by checking the dimensionality of the value vs the logp. We use this logic in some places already:

https://github.com/pymc-devs/pymc/blob/f67ff8bc22fcc72ead96eeeaef04173ed53650cb/pymc/logprob/transforms.py#L432-L437

https://github.com/pymc-devs/pymc/blob/f67ff8bc22fcc72ead96eeeaef04173ed53650cb/pymc/logprob/tensor.py#L185-L189

Which makes me worry whether the probability of a transformed broadcasted variable may be invalid because the "Jacobian" term is going to be counted multiple times?

File: pymc/logprob/shape.py (hunk 3 above)
Comment (ricardoV94) [reply pymc/logprob/shape.py:63]: Yes

File: pymc/logprob/shape.py (hunk 4 above)
Comment (ricardoV94) [reply pymc/logprob/shape.py]: No, this method is what guides the search for measurable variables. Quoting myself:

> Instead of eager bottom-up conversion to measurable variables in the IR rewrites, we only convert nodes whose outputs were marked as "needs_measuring". This is achieved with the new `PreserveRVMappings.request_measurable method`.

The `assign_custom_measurable_outputs` is completely gone.

File: pymc/logprob/shape.py (hunk 5 above)
Comment (ricardoV94) [reply pymc/logprob/shape.py]: We check if we can recreate the original broadcasted value by selecting the first entries and rebroadcasting. If we cannot, the graph couldn't have arise from broadcasting.

In the simplest case:
```python
import numpy as np

a = np.array([0, 0, 0, 0, 0])
assert np.all(a == np.broadcast_to(a[0], a.shape))  # Checks out

a = np.array([0, 1, 2, 3, 4])
assert np.all(a == np.broadcast_to(a[0], a.shape))  # Raises
```

File: tests/logprob/test_transforms.py (hunk 6 above)
Comment (ricardoV94) [reply tests/logprob/test_transforms.py]: No. It should work for anything that is explicitly under `BroadcastTo` or implicitly in one of the supported binary Elemwise (Add/ Mul / Pow...)

File: pymc/logprob/shape.py (hunk 4 above)
Comment (larryshamalama) [reply pymc/logprob/shape.py]: I understand, I'm still getting used to these newer changes

File: pymc/logprob/shape.py (hunk 5 above)
Comment (larryshamalama) [reply pymc/logprob/shape.py]: Thanks for clarifying, the example here helps

File: pymc/logprob/shape.py (hunk 1 above)
Comment (ricardoV94) [reply pymc/logprob/shape.py:34]: You raised a very good point, which makes me wonder to what extent https://github.com/pymc-devs/pymc/pull/6797 is correct in general? 

For instance, if you scale a `3-vector Dirichlet` you shouldn't count the Jacobian 3 times, because one of the entries is redundant.

Do we need to propagate information about over-determined elements in multi-dimensional RVs?

Comment (ricardoV94) [reply pymc/logprob/shape.py:34]: The first part of this answer suggests you count it 3 times indeed: https://stats.stackexchange.com/a/487538

I'm surprised :D

Edit: As seen below, that answer is wrong

Comment (ricardoV94) [reply pymc/logprob/shape.py:34]: This I think says something else and correct? https://upcommons.upc.edu/bitstream/handle/2117/366723/p20-CoDaWork2011.pdf?sequence=1&isAllowed=y

I think these should match:
```python
import pymc as pm
import numpy as np

x = 0.75
print(
    pm.logp(pm.Beta.dist(5, 9), x).eval(),
    pm.logp(pm.Dirichlet.dist([5, 9]), [x, 1-x]).eval(),
)  # -3.471576058736023 -3.471576058736023

print(
    pm.logp(2 * pm.Beta.dist(5, 9), 2 * x).eval(),
    pm.logp(2 * pm.Dirichlet.dist([5, 9]), 2*np.array([x, 1-x])).eval(),
)  # -4.164723239295968 -4.857870419855914

print(
    pm.logp(2 * pm.Beta.dist(5, 9), 2 * x).eval(),
    (pm.logp(pm.Dirichlet.dist([5, 9]), ([x, 1-x])) - np.log(2)).eval(),
)  # -4.164723239295968 -4.164723239295968
```

Comment (larryshamalama) [reply pymc/logprob/shape.py:34]: > Once we have the meta-info, the Mixture will unambiguously know what kind of measurable variable it is dealing with. In the case of MeasurableBroadcasting, for example, the `ndim_supp` will have to be at least as large as the number of broadcasted dims (which means we should collapse that logp dimension instead of leaving it as we were doing now!).

This makes sense! Would you say that it's better to wait for #6360?

> The first part of this answer suggests you count it 3 times indeed: https://stats.stackexchange.com/a/487538
> 
> I'm surprised :D

I'm not sure if I fully follow 😅 Nonetheless, I'm glad that this question raised some interesting concerns

**Review by jessegrabowski — APPROVED** [review]
